aggregator/src/aggregator/aggregation_job_writer.rs

```diff
@@ -0,0 +1,471 @@
+//! In-memory accumulation of aggregation job (& report aggregation) updates.
+
+use crate::{aggregator::query_type::CollectableQueryType, Operation};
+use futures::{future::try_join_all, TryFutureExt};
+use janus_aggregator_core::{
+    datastore::{
+        models::{
+            AggregationJob, AggregationJobState, Batch, BatchState, CollectionJobState,
+            ReportAggregation, ReportAggregationState,
+        },
+        Error, Transaction,
+    },
+    task::Task,
+};
+use janus_core::time::Clock;
+use janus_messages::{AggregationJobId, ReportId, ReportShareError};
+use prio::{codec::Encode, vdaf};
+use std::{
+    borrow::Cow,
+    collections::{HashMap, HashSet},
+    hash::Hash,
+    sync::{Arc, Mutex},
+};
+use tokio::try_join;
+
+/// AggregationJobWriter contains the logic used to write aggregation jobs, both initially &
+/// on updates. It is used only by the Leader.
+pub struct AggregationJobWriter<
```

Comment:
Unfortunately, we can't make the receiver `self` rather than `&self` as we really do need to call `write` repeatedly on transaction retries (and some use cases, e.g. the aggregation job driver, will make their `put`/`update` calls outside of the transaction -- so we really do want to reuse the same `AggregationJobWriter` state on these repeated calls, as we can't refactor to create a new `AggregationJobWriter` each time without moving some outside-of-transaction work into a transaction).

I updated the `write` docs to note that `write` doesn't drop any state on a write.

Reply:
Good enough, we don't have to get all that fancy with internal-facing interfaces.